lib/src/metta/runner/stdlib.rs

```diff
@@ -708,6 +708,65 @@ impl Grounded for PrintlnOp {
     }
 }
 
+// Implement trace! built-in.
+//
+// It is equivalent to Idris or Haskell Trace, that is, it prints a
+// message to stderr and pass a value along.
+//
+// For instance
+// ```
+// !(trace! "Here?" 42)
+// ```
+// prints to stderr
+// ```
+// Here?
+// ```
+// and returns
+// ```
+// [42]
+// ```
+//
+// Note that the first argument does not need to be a string, which
+// makes `trace!` actually quite capable on its own.  For instance
+// ```
+// !(trace! ("Hello world!" (if True A B) 1 2 3) 42)
+// ```
+// prints to stderr
+// ```
+// (Hello world! A 1 2 3)
+// ```
+// and returns
+// ```
+// [42]
+// ```
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct TraceOp {}
+
+impl Display for TraceOp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "trace!")
+    }
+}
+
+impl Grounded for TraceOp {
+    fn type_(&self) -> Atom {
+        Atom::expr([ARROW_SYMBOL, ATOM_TYPE_UNDEFINED, Atom::var("a"), Atom::var("a")])
+    }
+
+    fn execute(&self, args: &mut Vec<Atom>) -> Result<Vec<Atom>, ExecError> {
+        let arg_error = || ExecError::from("trace! expects two atoms as arguments");
+        let msg = args.get(0).ok_or_else(arg_error)?;
+        let val = args.get(1).ok_or_else(arg_error)?.clone();
```

Comment:
Using `pop()` instead of `get()` eliminates cloning, but arguments need to be retrieved in reversed order:
```suggestion
        let val = args.pop().ok_or_else(arg_error)?;
        let msg = args.pop().ok_or_else(arg_error)?;
```